brave-apache-http-interceptors/src/main/java/brave/httpclient4/HttpClient4Instrumentation.java

```diff
@@ -0,0 +1,126 @@
+package brave.httpclient4;
+
+import brave.ClientHandler;
+import brave.Span;
+import brave.Tracer;
+import brave.parser.Parser;
+import brave.parser.TagsParser;
+import brave.propagation.Propagation;
+import brave.propagation.TraceContext;
+import java.io.IOException;
+import java.net.URI;
+import org.apache.http.HttpMessage;
+import org.apache.http.HttpRequestInterceptor;
+import org.apache.http.HttpResponseInterceptor;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.client.protocol.HttpClientContext;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import zipkin.TraceKeys;
+
+import static com.github.kristofa.brave.internal.Util.checkNotNull;
+
+public final class HttpClient4Instrumentation {
```

Comment:
we've been doing this wrong :) if we want folks to use tracer.currentSpan we need to implement HttpProcessor so that we can scope the span before any of the interceptors fire.

that way, people can write an interceptor that uses the current span without needing to know a magic attribute. This is particularly important for slf4j or log4j2 mdc logging. If we set the span ID before any interceptors go, anything that happens will end up with the same ids in the log. The only way I can see how to do this is implementing a TracingHttpProcessor

cc @shakuzen